### PR TITLE
Epicodim3

### PR DESCRIPTION
**Pull request by LenSpek**
My implementation of the Epileptor Codim 3 model by Marisa. 
(Ignore the first 3 commits, they were added by accident)

### CODE REVIEW

File: tvb/simulator/models/epileptorcodim3.py [hunk 1]
```diff
@@ -0,0 +1,595 @@
...
+from .base import Model, LOG, numpy, basic, arrays
+
+
+class EpileptorCodim3(Model):
+    r"""
+    .. [Saggioetal_2017] Saggio ML, Spiegler A, Bernard C, Jirsa VK. *Fast–Slow Bursters in the Unfolding of a
+        High Codimension Singularity and the Ultra-slow Transitions of Classes.*
+        Journal of Mathematical Neuroscience. 2017;7:7. doi:10.1186/s13408-017-0050-8.
+
+    .. The Epileptor codim 3 model is a neural mass model which contains two subsystems acting at different timescales.
```
Comment (maedoc) [inline tvb/simulator/models/epileptorcodim3.py]: please reflow to 80 columns, in Vim this is `gqip` but your editor may be different. or I can do it for you.

File: tvb/simulator/models/epileptorcodim3.py [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+
+        A = numpy.array([self.mu2_start[0], -self.mu1_start[0], self.nu_start[0]])
+        B = numpy.array([self.mu2_stop[0], -self.mu1_stop[0], self.nu_stop[0]])
+
+        self.E = A / numpy.linalg.norm(A)
+        self.F = numpy.cross(numpy.cross(A, B), A)
+        self.F = self.F / numpy.linalg.norm(self.F)
+
+
+class EpileptorCodim3_slowmod(Model):
```
Comment (maedoc) [inline tvb/simulator/models/epileptorcodim3.py]: Is this a completely different model?

**Comment by maedoc** [comment]
@LenSpek looking good; I have a few requests: 

- [ ] keep the `dfun`s as you have them but try to add numba versions, as [in the epileptor 6d model](https://github.com/the-virtual-brain/tvb-library/blob/trunk/tvb/simulator/models/epileptor.py#L37), and [use it as the dfun](https://github.com/the-virtual-brain/tvb-library/blob/trunk/tvb/simulator/models/epileptor.py#L356).  please test & eventually commit to this branch/PR
- [ ] please provide some reproductions of figures from the paper either as a Python script or prefereably as a Jupyter notebook

File: tvb/simulator/models/epileptorcodim3.py (hunk 2 above)
Comment (LenSpek) [reply tvb/simulator/models/epileptorcodim3.py]: This model is similar, it is a different implementation. It allows for multiple classes of bursters to be visited by a simulation. The model dynamics are essentially the same. However there are some differences in the input of the system and some extra dynamics to allow the transition of classes

File: tvb/simulator/models/epileptorcodim3.py (hunk 1 above)
Comment (LenSpek) [reply tvb/simulator/models/epileptorcodim3.py]: I will adjust this

**Review by maedoc — COMMENTED** [review]
Looking good! 👍🏼
Careful though, if you implement Numba dfun for second model, you should change the names to be unique .

**Comment by maedoc** [comment]
How much time do you have left to work in this model, btw?

**Comment by LenSpek** [comment]
I have time till the end of January. However if there are a few minor things, I will be able to address these after my internship has ended

**Comment by maedoc** [comment]
thanks and sorry this took so long.

**Comment by liadomide** [comment]
After this recent merge, one unit-test fails: 
```
tvb-library/tvb/tests/simulator/simulator_test.py: TestSimulator.test_simulator_region

Error Message
TypeError: ufunc '_numba_dfun' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
....
../scientific_library/tvb/simulator/models/epileptorcodim3.py:286: TypeError

```
Can you please check if the test was wrongly written, or the new model is missing some API implicit requirement ?

**Comment by maedoc** [comment]
The test probably assumes all parameters are floats while some parameters in this model are integers, and it's unsafe to convert a float to an integer.

File: tvb/simulator/models/epileptorcodim3.py [hunk 3]
```diff
@@ -0,0 +1,720 @@
...
+
+        self.E = A / numpy.linalg.norm(A)
+        self.F = numpy.cross(numpy.cross(A, B), A)
+        self.F = self.F / numpy.linalg.norm(self.F)
+
+    def dfun(self, state_variables, coupling, local_coupling=0.0):
+        """"The dfun using numba for speed"""
+        state_variables_ = state_variables.reshape(state_variables.shape[:-1]).T
+        coupling_ = coupling.reshape(coupling.shape[:-1]).T
+        derivative = _numba_dfun(state_variables_, coupling_, self.E[0], self.E[1], self.E[2], self.F[0], self.F[1],
```
Comment (maedoc) [inline tvb/simulator/models/epileptorcodim3.py:285]: @LenSpek can you set a breakpoint here and double check that the types of those non-float parameters being passed to the numba dfun match or if not, cast their values correclty?

Comment (LenSpek) [reply tvb/simulator/models/epileptorcodim3.py:285]: I double-checked and the variables in question (self.modification, self.N) have dtype int32.

Comment (maedoc) [reply tvb/simulator/models/epileptorcodim3.py:285]: is that the case for defaults as well (when you don't specify any model parameters)?  this test runs your model as if you made a simulation without specifying any parameters, only defaults

**Comment by LenSpek** [comment]
I think that I found what was triggering the tests: Modification is specified by default as a BoolArray, which TVB converts to an int32. I will push a change to address this, so the test works. 
(BTW, when I run the TVB tests, I don't get this error)

**Comment by liadomide** [comment]
If i debug in my env, the 2 params which are expected to be int32 are in fact int64.
I get the same error both in unit-test and in the ipython notebook that came with these new models.
I suspect this is the default of the installation.

**Comment by LenSpek** [comment]
I find that strange as I use the TVB classes from array.py and there the class IntegerArray is defined as:
```python
class IntegerArray(BaseArray):
    _ui_name = "Array of integers"
    dtype = basic.DType(default=numpy.int32)
```

**Comment by liadomide** [comment]
Could we use dtype **int_** ?
This would become **int32** on Windows OS and **int64** on Unix systems.

I tested on _Anaconda x64_ with _Python 2.7.14_ and _numpy 1.13_ on Linux, Mac and Windows. If we replace the current _int32_ with _int__, then all 3 systems are ok. Otherwise, with the current code only Windows is fine, the others are not working.

**Comment by maedoc** [comment]
@liadomide  can you paste a `git diff` of your `int_` change, so it's clear where the change is made?

**Comment by maedoc** [comment]
Numba [supports](http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html) the dtypes in question, and I would prefer platform independence if possible. (`int_` corresponds to C `long`)

**Comment by liadomide** [comment]
[diff.txt](https://github.com/the-virtual-brain/tvb-library/files/1645879/diff.txt)

**Comment by liadomide** [comment]
`int_` is the default Integer type for Numpy, that is why I suggest it. 
Seems to work well with numba `int_`
https://docs.scipy.org/doc/numpy-1.13.0/user/basics.types.html

**Comment by maedoc** [comment]
OK then, `int_` it is.

**Comment by liadomide** [comment]
Good. I will then apply this change soon.

**Pull request merged**